File: clients/python/examples/subscribe.py
#!/usr/bin/env python3
"""
Example: Subscribing to messages from NexusD with auto-reconnection and gap detection
"""

import json
import signal
import sys
from nexusd_client import (
    NexusdClient,
    ReconnectionPolicy,
    GapRecoveryMode,
    ReconnectionState,
)

# Handle Ctrl+C gracefully
def signal_handler(sig, frame):
    print("\nShutting down...")
    sys.exit(0)

signal.signal(signal.SIGINT, signal_handler)


def main():
    # Configure reconnection policy
    policy = ReconnectionPolicy(
        max_retries=0,  # 0 = infinite retries
        initial_delay_ms=1000,
        max_delay_ms=30000,
        exponential_backoff=True,
        jitter_factor=0.1,
    )

    # Connect to the daemon with resilience settings
    with NexusdClient(
        "localhost:5672",
        reconnection_policy=policy,
        gap_recovery_mode=GapRecoveryMode.REPLAY_BUFFER,
    ) as client:
        print("Connected to NexusD")
        print("Subscribing to sensors/* topics...")
        print("Auto-reconnect enabled with gap recovery")
        print("Press Ctrl+C to stop\n")

        # Register connection state callback
        def on_state_change(state: ReconnectionState, error: Exception | None):
            if state == ReconnectionState.CONNECTED:
                print("[Connection] Connected to daemon")
            elif state == ReconnectionState.RECONNECTING:
                print(f"[Connection] Reconnecting... ({error})")
            elif state == ReconnectionState.DISCONNECTED:
                print("[Connection] Disconnected")
            elif state == ReconnectionState.FAILED:
                print(f"[Connection] Failed: {error}")

        client.on_connection_state_change(on_state_change)

        # Register gap detection callback
        def on_gap(expected_seq: int, actual_seq: int, topic: str):
            gap = actual_seq - expected_seq
            print(f"[Gap] Detected {gap} missed messages on {topic}")
            print(f"       Expected seq {expected_seq}, got {actual_seq}")

        client.on_gap_detected(on_gap)

        # Subscribe to multiple topics
        topics = ["sensors/temperature", "sensors/humidity", "config/settings"]

        for message in client.subscribe(topics, receive_retained=True):
            # Parse JSON payload if applicable
            try:
                if message.content_type == "application/json":
                    payload = json.loads(message.payload.decode())
                else:
                    payload = message.payload.decode()
            except:
                payload = message.payload

            # Print message details with sequence number
            retained_flag = " [RETAINED]" if message.is_retained else ""
            print(f"Topic: {message.topic}{retained_flag}")
            print(f"  Payload: {payload}")
            print(f"  Sequence: {message.sequence_number}")
            print(f"  Message ID: {message.message_id}")
            print(f"  Timestamp: {message.timestamp_ms}")
            print()


if __name__ == "__main__":
    main()

File: clients/node/examples/subscribe.ts
/**
 * Example: Subscribing to messages with NexusD Node.js client
 * 
 * Demonstrates auto-reconnection with exponential backoff and gap detection.
 */

import { NexusdClient, GapRecoveryMode, ReconnectionState } from '../src';

async function main() {
    // Connect to daemon with resilience settings
    const client = new NexusdClient('localhost:5672', {
        reconnection: {
            maxRetries: 0,  // 0 = infinite retries
            initialDelayMs: 1000,
            maxDelayMs: 30000,
            exponentialBackoff: true,
            jitterFactor: 0.1,
        },
        gapRecoveryMode: GapRecoveryMode.ReplayBuffer,
        debug: true,  // Enable structured logging
    });

    console.log('Connected to NexusD');
    console.log('Subscribing to sensors/* topics...');
    console.log('Auto-reconnect enabled with gap recovery');
    console.log('Press Ctrl+C to stop\n');

    // Register connection state callback
    client.onConnectionStateChange((state, error) => {
        switch (state) {
            case ReconnectionState.Connected:
                console.log('[Connection] Connected to daemon');
                break;
            case ReconnectionState.Reconnecting:
                console.log(`[Connection] Reconnecting... (${error?.message})`);
                break;
            case ReconnectionState.Disconnected:
                console.log('[Connection] Disconnected');
                break;
            case ReconnectionState.Failed:
                console.log(`[Connection] Failed: ${error?.message}`);
                break;
        }
    });

    // Register gap detection callback
    client.onGapDetected((expected, actual, topic) => {
        const gap = actual - expected;
        console.log(`[Gap] Detected ${gap} missed messages on ${topic}`);
        console.log(`       Expected seq ${expected}, got ${actual}`);
    });

    // Handle shutdown
    process.on('SIGINT', () => {
        console.log('\nShutting down...');
        client.close();
        process.exit(0);
    });

    try {
        // Subscribe to multiple topics with gap recovery
        const topics = ['sensors/temperature', 'sensors/humidity', 'config/settings'];

        for await (const message of client.subscribe(topics, { receiveRetained: true })) {
            const retainedFlag = message.isRetained ? ' [RETAINED]' : '';
            console.log(`Topic: ${message.topic}${retainedFlag}`);

            try {
                if (message.contentType === 'application/json') {
                    console.log(`  Payload: ${JSON.stringify(message.payloadAsJson())}`);
                } else {
                    console.log(`  Payload: ${message.payloadAsString()}`);
                }
            } catch {
                console.log(`  Payload: ${message.payload.toString('hex')}`);
            }

            console.log(`  Sequence: ${message.sequenceNumber}`);
            console.log(`  Message ID: ${message.messageId}`);
            console.log(`  Timestamp: ${message.timestampMs}`);
            console.log();
        }
    } finally {
        client.close();
    }
}

main().catch(console.error);
